packages/teuchos/comm/src/Teuchos_StackedTimer.cpp

```diff
@@ -67,33 +66,90 @@ StackedTimer::merge(Teuchos::RCP<const Teuchos::Comm<int> > comm){
 }
 
 void
-StackedTimer::collectRemoteData(Teuchos::RCP<const Teuchos::Comm<int> > comm) {
-  int comm_size = size(*comm);
+StackedTimer::collectRemoteData(Teuchos::RCP<const Teuchos::Comm<int> > comm, const OutputOptions &options) {
   int comm_rank = rank(*comm);
-  Array<BaseTimer::TimeInfo> local_time_data;
-  if (comm_rank == 0) {
-    time_data_.resize(comm_size);
-    for (int r=0; r<comm_size; ++r)
-      time_data_[r].resize(flat_names_.size());
-  } else {
-    local_time_data.resize(flat_names_.size());
+
+  // allocate everything
+  int num_names = flat_names_.size();
+  sum_.resize(num_names);
+  count_.resize(num_names);
+  active_.resize(num_names);
+
+  if (options.output_minmax || options.output_histogram) {
+    min_.resize(num_names);
+    max_.resize(num_names);
+    if ( options.output_minmax )
+      sum_sq_.resize(num_names);
   }
 
-  for (unsigned i=0; i<flat_names_.size(); ++i) {
+
+  if (options.output_histogram && comm_rank == 0 ) {
+    hist_.resize(options.num_histogram);
+    for (int i=0;i<options.num_histogram; ++i)
+      hist_[i].resize(num_names);
+  }
+
+  // Temp data
+  Array<double> time(num_names);
+  Array<unsigned long> count(num_names);
+  Array<unsigned long long> updates;
+  if (options.output_total_updates)
+    updates.resize(num_names);
+  Array<int> used(num_names);
+  Array<int> bins;
+
+  if (options.output_histogram)
+    bins.resize(num_names);
+
+  // set initial values
+  for (int i=0;i<num_names; ++i) {
     auto t = timer_.findTimer(flat_names_[i]);
-    if ( comm_rank == 0 )
-      time_data_[0][i] = t;
-    else
-      local_time_data[i] = t;
+      time[i] = t.time;
+      count[i] = t.count;
+      used[i] = t.count==0? 0:1;
+      if (options.output_total_updates)
+        updates[i] = t.updates;
+  }
+
+  // Now reduce the data
+  reduce<int, double>(&time[0], &sum_[0], num_names, REDUCE_SUM, 0, *comm);
```

Comment:
It would be better to use `x.getRawPtr()` instead of `&x[0]`, since that won't throw any illegal access errors if the thing is empty.